Documentation/ceph-encryption.md

```diff
@@ -0,0 +1,225 @@
+        resources:
+          requests:
+            storage: 10Gi
+  cephVersion:
+    image: quay.io/ceph/ceph:v16.2.6
+    allowUnsupported: false
+  mgr:
+    count: 1
+  storage:
+    storageClassDeviceSets:
+      - name: set1
+        count: 1
+        portable: false
+        tuneDeviceClass: true
+        tuneFastDeviceClass: false
+        encrypted: true    # ==> enable encryption
+```
+
+## Customer requirements & encryption at rest
+There are scenarios when the managers of the Ceph cluster are different than the
+consumers of the cluster. In those scenarios, encryption at rest can be a legal
+and compliance requirement, that includes key rotation support and the ability
+for customers to bring their own encryption keys.
+
+This introduces new feature requirements into rook/ceph:
+1. Support for key rotation.
+2. Support for BYOK for encryption.
+
+## Key Rotation
+As of today, neither Ceph nor Rook support a flag or easy mechanism to rotate a
+Ceph key. Key rotation is feasible through generating a new key and adding it to
+the LUKS header using cryptsetup. There are 2 different approaches to this:
+-  Updating the LUKS key: LUKS header can be updated to change the key, ex:
+   `cryptsetup luksChangeKey --key-file <new key file>` however this doesn't
+   change the `volume key` which is the master key used for encryption.
+-  Updating the master key requires reencryption and can be done through
+   `cryptsetup-reencrypt` which reencrypts data on LUKS device in-place. During
+   reencryption process the LUKS device is marked unavailable.
+   `cryptsetup-reencrypt --key-file <new key file>`.
+
+References:
+- https://man7.org/linux/man-pages/man8/cryptsetup-reencrypt.8.html
+- https://man7.org/linux/man-pages/man8/cryptsetup.8.html#LUKS_EXTENSION
+
+To be able to fulfill key rotation requirements, changes are required to either
+Ceph or Rook depending on the Rook Ceph cluster deployment type:
+
+1.  Host based clusters:
+
+    a.  Ceph: Key rotation will need to be enabled in Ceph and a flag exposed to
+    enabled it. For rotation frequency, it can default to 90 days which is
+    commonly used.
+
+    b.  Rook: Add support to expose and pass the required flags to enable key
+    rotation.
+
+2.  PVC based clusters: Since keys are managed by Rook, it is easier to add such
+    capability into rook. This can be added as an extra flag to the cluster crd
+    `rotateEncryptionKey: true`. Once specified, rook can automatically
+    regenerate a new key on certain frequency (can default to 90days) and update
+    the LUKS header as needed.
+
+
+## BYOK Support in Rook
+
+Customers may be already using existing key management systems to manage their
+secrets or relying on enterprise grade cloud providers for key management. Thus
+the proposal is to leverage  [secrets store
+csi](https://github.com/kubernetes-sigs/secrets-store-csi-driver) for mounting
+of secrets from external secret stores into the OSDs.
```

Comment:
If I understand the project correctly, it will also be valuable outside of the BYOK use case, right? If we can get any key stored in a KMS mounted as a volume.

Reply:
That's correct. I was thinking of using it initially for BYOK. We can later expand the scope to other rook secrets too, where it makes sense.

Reply:
Right, using it for the existing encryption component will require a lot of work. So let's target it for BYOK. However, I wonder how this will work in terms of dependency. We are expecting https://github.com/kubernetes-sigs/secrets-store-csi-driver to be deployed and available right?

Reply:
@leseb If we are able to implement this using a more generic solution based on the secrets-store-csi-driver, would we want to deprecate the existing integrations after some amount of time? It seems to me that using secrets-store-csi-driver would be more flexible and allow us to maintain less code in the long term.

Reply:
It's certainly something we need to consider. However, we need to evaluate if we have backends and feature parity with what we have today.